src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,373 @@
+
+        push_nonvol_reg r9
+        push_nonvol_reg r11
+        push_nonvol_reg rbp
+        alloc_stack     UM2MThunk_WrapperHelper_FIXED_STACK_ALLOC_SIZE
+        set_frame       rbp, UM2MThunk_WrapperHelper_FRAME_OFFSET
+        END_PROLOGUE
+
+        //
+        // We are in cooperative mode and in the correct domain. 
+        // The host has also been notified that we've entered the 
+        // runtime.  All we have left to do is to copy the stack, 
+        // setup the register args and then call the managed target
+        //
+
+        test            rsi, rsi
+        jg              LOCAL_LABEL(UM2MThunk_WrapperHelper_CopyStackArgs)
+
+LOCAL_LABEL(UM2MThunk_WrapperHelper_ArgumentsSetup):
+        mov             METHODDESC_REGISTER, r9
+
+        mov             r11, rdi                // rsi <- pThunkArgs
+        mov             rdi, [r11 +  0h]
+        mov             rsi, [r11 +  8h]
+        mov             rdx, [r11 + 10h]
+        mov             rcx, [r11 + 18h]
+        mov             r8, [r11 + 20h]
+        mov             r9, [r11 + 28h]
+
+        movdqa          xmm0, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h]
+        movdqa          xmm1, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h]
+        movdqa          xmm2, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 20h]
+        movdqa          xmm3, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 30h]
+        movdqa          xmm4, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 40h]
+        movdqa          xmm5, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 50h]
+        movdqa          xmm6, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 60h]
+        movdqa          xmm7, [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 70h]
+
+        mov             rax, [METHODDESC_REGISTER + OFFSETOF__UMEntryThunk__m_pUMThunkMarshInfo]      // rax <- UMThunkMarshInfo*
+        mov             rax, [rax + OFFSETOF__UMThunkMarshInfo__m_pILStub]                            // rax <- Stub*
+        call            rax
+
+        // make sure we don't trash the return value
+        mov             [r11 + 0h], rax
+        movdqa          [r11 + UMThunkStubAMD64_XMM_SAVE_OFFSET - UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h], xmm0
+
+        lea             rsp, [rbp - UM2MThunk_WrapperHelper_FRAME_OFFSET + UM2MThunk_WrapperHelper_FIXED_STACK_ALLOC_SIZE]
+        pop             rbp
+        pop             r11
+        pop             r9
+        ret
+        
+
+LOCAL_LABEL(UM2MThunk_WrapperHelper_CopyStackArgs):
+        // rdx = cbStackArgs (with 20h for register args subtracted out already)
```

Comment:
We need to also preserve rcx here, so I'll use rdx as a scratch for rdi.  It looks like r10 can be used for rcx?

*note* spelling autocorrect for registers sucks